de.dlr.sc.virsat.model.extension.fdir/src/de/dlr/sc/virsat/model/extension/fdir/modularizer/Module.java

```diff
@@ -67,20 +88,102 @@ public FaultTreeNode getRootNode() {
-	 * @return is dynamic
+	 * Get the root node of the copied module
+	 * @return the root node of the copied module
 	 */
-	public boolean isDynamic() {
-		return this.moduleType == ModuleType.DYNAMIC;
+	public FaultTreeNode getRootNodeCopy() {
+		return this.moduleRootCopy;
+	}
+	
+	/**
+	 * Get the map which maps original fault tree nodes to the copy fault tree nodes
+	 * @return the map
+	 */
+	public Map<FaultTreeNode, FaultTreeNode> getMapOriginalToCopy() {
+		return this.mapOriginalToCopy;
+	}
+	
+	
+	/**
+	 * Returns true if module is nondeterministic, false otherwise
+	 * @return is nondeterministic
+	 */
+	public boolean isNondeterministic() {
+		return this.moduleType == ModuleType.NONDETERMINISTIC;
 	}
 	
 	/**
 	 * Trim the module from the tree.
 	 */
 	public void harvestFromFaultTree() {
-		for (FaultTreeNodePlus node : this.moduleNodes) {
-			node.harvestFromFaultTree();
+		this.moduleNodes.stream().forEach(node -> node.harvestFromFaultTree());
+	}
+	
+	/**
+	 * Create the fault tree copy that is required for conversion to markov automata, with only the edges in the module
+	 */
+	public void constructFaultTreeCopy() {
+		FaultTreeHelper fthelp = new FaultTreeHelper(this.moduleNodes.get(0).getFaultTreeNode().getConcept());
+		Stack<FaultTreeNode> dfsStack = new Stack<FaultTreeNode>();
+		this.mapOriginalToCopy = new HashMap<FaultTreeNode, FaultTreeNode>();
+		
+		FaultTreeNode originalRoot = this.moduleNodes.get(0).getFaultTreeNode().getFault().getFaultTree().getRoot();
+		FaultTreeNode rootCopy = fthelp.copyFaultTreeNode(originalRoot, null);
+		rootCopy.setName(originalRoot.getName());
+
+		List<FaultTreeNode> sparesInOriginalFaultTree = fthelp.getAllSpareNodes(originalRoot.getFault());
+		mapOriginalToCopy.put(originalRoot, rootCopy);
+		dfsStack.push(originalRoot);
+		
+		while (!dfsStack.isEmpty()) {
+			FaultTreeNode curr = dfsStack.pop();
+			FaultTreeNode currCopy = mapOriginalToCopy.get(curr);
+			
+			if (curr.equals(moduleRoot.getFaultTreeNode())) {
+				this.moduleRootCopy = currCopy;
+			}
+			
+			List<FaultTreeNode> children = fthelp.getAllChildren(curr, curr.getFault().getFaultTree());
+			for (FaultTreeNode child : children) {
+				FaultTreeNode childCopy;
+				if (mapOriginalToCopy.get(child) == null) {
+					childCopy = fthelp.copyFaultTreeNode(child, currCopy.getFault());
+					mapOriginalToCopy.put(child, childCopy);
+					for (int i = 0; i < childCopy.getFault().getBasicEvents().size(); ++i) {
+						BasicEvent oldBasicEvent = child.getFault().getBasicEvents().get(i);
+						BasicEvent newBasicEvent = childCopy.getFault().getBasicEvents().get(i);
+						mapOriginalToCopy.put(oldBasicEvent, newBasicEvent);
+					}
+				} else {
+					childCopy = mapOriginalToCopy.get(child);
+				}
+				
+				boolean moduleContainsCurrAndChild = this.containsFaultTreeNode(curr) && this.containsFaultTreeNode(child);
+				if (moduleContainsCurrAndChild
+						&& child.getFaultTreeNodeType() != FaultTreeNodeType.BASIC_EVENT) {

```

Comment:
I think `!child.getFaultTreeNodeType.equals(FaultTreeNodeType.BASIC_EVENT)`  would be better to read